pyro/distributions/torch_patch.py

```diff
@@ -33,6 +33,49 @@ def decorator(new_fn):
     return decorator
 
 
+# See: tests.distributions.transforms.test_transforms.py::test_transform_backward
+# TODO: should be moved to PyTorch if this is a reasonable pattern
+@patch_dependency('torch.distributions.transforms.Transform.__call__')
+def _Transform__call__(self, x):
+    if self._cache_size == 0:
+        return self._call(x)
+    x_old, y_old = self._cached_x_y
+    if x is x_old:
+        return y_old
+    y = self._call(x)
+
+    if self._cache_size > 0:
+        # unset cached value after call to backward
+        def _hook(_):
+            self._cached_x_y = None, None
+
+        y.register_hook(_hook)
```

Comment:
I believe this creates a cyclic reference (unless PyTorch uses weakrefs in hook logic).

@apaszke do you have recommendations for avoiding cyclic dependencies in hooks?

Reply:
Good catch. I think one way might be to capture the handle as `self._handle = y.register_hook(_hook)` and to remove it `self._handle.remove()` when a new value of `y` is provided. I'll have to test it out, but there might be a simpler solution.

Reply:
Maybe
```py
weakself = weakref.ref(self)
def _hook(_):
    self = weakself()
    if self is not None:
        self._cached_x_y = None, None
```

Reply:
That should work great!